dotnet-interactive/JupyterKernelSpecInstaller.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Interactive.App
+{
+    public class JupyterKernelSpecInstaller : IJupyterKernelSpecInstaller
+    {
+        private readonly JupyterKernelSpecModule _kernelSpecModule;
+
+        public JupyterKernelSpecInstaller()
+        {
+            _kernelSpecModule = new JupyterKernelSpecModule();
+        }
+        public async Task<KernelSpecInstallResult> InstallKernel(DirectoryInfo sourceDirectory, DirectoryInfo destination = null)
+        {
+
+            if (destination != null)
+            {
+                var (succeeded, message) = CopyKernelSpecFiles(sourceDirectory, destination);
+                return new KernelSpecInstallResult(succeeded, message);
+            }
+            else
+            {
+                try
+                {
+                    var result = await _kernelSpecModule.InstallKernel(sourceDirectory);
+                    if (result.ExitCode == 0)
+                    {
+                        return new KernelSpecInstallResult(true,
+                            string.Join('\n', result.Output.Concat(result.Error)));
+                    }
+                }
+                catch (Win32Exception w32e)
+                {
+                    // file not found when executing process
+                    if (!w32e.Source.Contains(typeof(System.Diagnostics.Process).FullName))
+                    {
+                        throw;
+                    }
+                }
+
+                var notAvailable = "The kernelspec module is not available.";
+
+                var location = GetDefaultDirectory();
+                
+                if (!location.Exists)
+                {
+                    return new  KernelSpecInstallResult(false, string.Join('\n', notAvailable, $"The kernelspec path ${location.FullName} does not exist."));
```

Comment:
Why not create the default path and then install there? As a user, this allows me to install `dotnet-interactive` first and then install `nteract`, right?

Reply:
Interesting idea. Do you think this makes things easier and/or increases the potential for things to go wrong?

(Also, just to be nitpicky, these paths aren't specific to nteract.)

Reply:
I think the flexibility is nice to have. I cannot think of how this can make things go wrong, but I'm not as familiar to the Jupyter scenarios like you guys, so I might be wrong.

Reply:
It's possible, though it seems pretty unlikely, that some Jupyter installer could fail if the directory is already present or not empty.